lib/src/pgn.dart

```diff
@@ -0,0 +1,831 @@
+        nags!.add(nag);
+        newNags = nags!.toList();
+      } else {
+        newNags.add(nag);
+      }
+    }
+    return PgnNodeData(
+        san: san,
+        startingComments: startingComments ?? const [],
+        comments: comment == null ? null : newComments,
+        nags: nag == null ? null : newNags);
+  }
+}
+
+class _TransformStack<T, U, C> {
+  final PgnNode<T> before;
+  final PgnNode<U> after;
+  final C ctx;
+
+  _TransformStack(this.before, this.after, this.ctx);
+}
+
+/// Parent Node containing list of child nodes (Does not contain any data)
+class PgnNode<T> {
+  /// Child nodes with PGN data
+  final List<PgnChildNode<T>> children = [];
+  PgnNode();
+
+  /// Implements a Iterable for the node and its children
+  ///
+  /// Used for only iterating the mainline
+  Iterable<T> mainline() sync* {
+    var node = this;
+    while (node.children.isNotEmpty) {
+      final child = node.children[0];
+      yield child.data;
+      node = child;
+    }
+  }
+
+  /// Function to walk thorugh each node and transform Node<T> tree into Node<U> tree
+  PgnNode<U> transform<V, U, C>(
+      PgnNode<V> node, C ctx, U? Function(C, V, int) f) {
+    final root = PgnNode<U>();
+    final stack = [_TransformStack(node, root, ctx)];
+
+    while (stack.isNotEmpty) {
+      final frame = stack.removeLast();
+      for (var childIdx = 0;
+          childIdx < frame.before.children.length;
+          childIdx++) {
+        final childBefore = frame.before.children[childIdx];
+        final data = f(ctx, childBefore.data, childIdx);
+        if (data != null) {
+          final childAfter = PgnChildNode(data);
+          frame.after.children.add(childAfter);
+          stack.add(_TransformStack(childBefore, childAfter, ctx));
+        }
+      }
+    }
+    return root;
+  }
+}
+
+/// Child Node with PGN data
+class PgnChildNode<T> extends PgnNode<T> {
+  /// PGN Data
+  T data;
+  PgnChildNode(this.data);
+}
+
+/// A game represented by headers and moves
+///
+/// Used to convert a chess game from or to a PGN
+@immutable
+class PgnGame<T> {
+  /// Headers of the game
+  final Headers headers;
+
+  /// Inital comments of the game
+  final List<String> comments;
+
+  /// Parant node containing the game
+  final PgnNode<T> moves;
+
+  /// Constant contructor of the class.
+  ///
+  /// Returns a new immutable Game
+  const PgnGame(
+      {required this.headers, required this.moves, required this.comments});
+
+  /// Default function headers of a PGN
+  static Headers defaultHeaders() => {
+        'Event': '?',
+        'Site': '?',
+        'Date': '????.??.??',
+        'Round': '?',
+        'White': '?',
+        'Black': '?',
+        'Result': '*'
+      };
+
+  /// Create empty headers of a PGN
+  static Headers emptyHeaders() => <String, String>{};
+
+  /// Parse a pgn and return a [PgnGame]
+  ///
+  /// Optinally provide a Function [initHeaders] to change the default headers
+  static PgnGame<PgnNodeData> parsePgn(String pgn,
```

Comment:
Can't use a factory constructor as the `PgnGame` take a generic type, so It can't return a `PgnGame<PgnNodeData>.` I can make it return by casting it, but then the linter complains about it.